File: app/src/main/java/dk/itu/moapd/virtualpiano/ui/PianoView.java
package dk.itu.moapd.virtualpiano.ui;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import dk.itu.moapd.virtualpiano.player.SoundPlayer;

public class PianoView extends View {

    private static final int sNumberOfKeys = 8;

    private final Paint mBlackColor;
    private final Paint mWhiteColor;
    private final Paint mGrayColor;

    private final ArrayList<PianoKey> mWhiteKeys = new ArrayList<>();
    private final ArrayList<PianoKey> mBlackKeys = new ArrayList<>();

    private int mPianoKeyWidth;
    private int mPianoKeyHeight;

    private final SoundPlayer mSoundPlayer;

    private final Handler mHandler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(@NonNull Message msg) {
            invalidate();
        }
    };

    public PianoView(Context context, AttributeSet attrs) {
        super(context, attrs);

        mBlackColor = new Paint();
        mBlackColor.setColor(Color.BLACK);

        mWhiteColor = new Paint();
        mWhiteColor.setColor(Color.WHITE);

        mGrayColor = new Paint();
        mGrayColor.setColor(Color.GRAY);
        mGrayColor.setStyle(Paint.Style.FILL);

        mSoundPlayer = new SoundPlayer(context);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldWidth, int oldHeight) {
        super.onSizeChanged(w, h, oldWidth, oldHeight);

        mPianoKeyWidth = w / sNumberOfKeys;
        mPianoKeyHeight = h;
        int count = sNumberOfKeys + 1;

        for (int i = 0; i < sNumberOfKeys; i++) {
            float left = i * mPianoKeyWidth;
            float right = (i + 1) * mPianoKeyWidth;

            if (i == sNumberOfKeys - 1)
                right = w;

            RectF rect = new RectF(left, 0, right, h);
            mWhiteKeys.add(new PianoKey(i + 1, rect));

            if (i != 0 && i != 3 && i != 7) {
                left = (float) (i - 1) * mPianoKeyWidth +
                        0.5f * mPianoKeyWidth +
                        0.25f * mPianoKeyWidth;
                right = (float) i * mPianoKeyWidth +
                        0.25f * mPianoKeyWidth;

                rect = new RectF(left, 0, right, 0.67f * h);
                mBlackKeys.add(new PianoKey(count++, rect));
            }
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        for (PianoKey key: mWhiteKeys)
            canvas.drawRect(key.getRectF(), key.isPressed() ? mGrayColor : mWhiteColor);

        for (int i = 1; i < sNumberOfKeys; i++)
            canvas.drawLine(i * mPianoKeyWidth, 0, i * mPianoKeyWidth,
                    mPianoKeyHeight, mBlackColor);

        for (PianoKey key: mBlackKeys)
            canvas.drawRect(key.getRectF(), key.isPressed() ? mGrayColor : mBlackColor);
    }

    @Override
    public boolean performClick() {
        super.performClick();
        return true;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        super.onTouchEvent(event);

        int action = event.getAction();
        boolean isPressed = (action == MotionEvent.ACTION_DOWN ||
                action == MotionEvent.ACTION_MOVE);

        if (action == MotionEvent.ACTION_UP)
            performClick();

        ArrayList<PianoKey> pressedKeys = new ArrayList<>();

        for (int i = 0; i < event.getPointerCount(); i++) {
            float x = event.getX(i);
            float y = event.getY(i);

            PianoKey key = keyForCoordinates(x, y);
            if (key != null) {
                key.setPressed(isPressed);
                pressedKeys.add(key);
            }
        }

        ArrayList<PianoKey> keyboard = new ArrayList<>(mWhiteKeys);
        keyboard.addAll(mBlackKeys);

        for (PianoKey key : keyboard) {
            if (pressedKeys.contains(key) && key.isPressed()) {
                if (mSoundPlayer.isNotPlaying(key.getSound())) {
                    mSoundPlayer.playNote(key.getSound());
                    invalidate();
                }
            } else {
                mSoundPlayer.stopNote(key.getSound());
                releasePianoKey(key);
            }
        }

        return true;
    }

    private PianoKey keyForCoordinates(float x, float y) {
        for (PianoKey key : mBlackKeys)
            if (key.getRectF().contains(x, y))
                return key;

        for (PianoKey key : mWhiteKeys)
            if (key.getRectF().contains(x, y))
                return key;

        return null;
    }

    private void releasePianoKey(final PianoKey key) {
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                key.setPressed(false);
                mHandler.sendEmptyMessage(0);
            }
        }, 100);
    }

}

File: app/src/main/java/dk/itu/moapd/virtualpiano/ui/PianoView.kt
package dk.itu.moapd.virtualpiano.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import dk.itu.moapd.virtualpiano.player.SoundPlayer

class PianoView(context: Context?, attrs: AttributeSet?) : View(context, attrs) {

    private val soundPlayer = SoundPlayer(context!!)

    private val blackColor = Paint().apply { color = Color.BLACK }
    private val whiteColor = Paint().apply { color = Color.WHITE }
    private val grayColor = Paint().apply {
        color = Color.GRAY
        style = Paint.Style.FILL
    }

    private val whiteKeys: ArrayList<PianoKey> = ArrayList()
    private val blackKeys: ArrayList<PianoKey> = ArrayList()

    private var keyWidth: Float = 0f
    private var keyHeight: Float = 0f

    private val mHandler: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            invalidate()
        }
    }

    companion object {
        private const val numberOfKeys = 8
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)

        keyWidth = (w / numberOfKeys).toFloat()
        keyHeight = h.toFloat()
        var count = numberOfKeys + 1

        for (i in 0 until numberOfKeys) {
            var left = i * keyWidth
            var right = (i + 1) * keyWidth

            if (i == numberOfKeys - 1)
                right = w.toFloat()

            var rect = RectF(left, 0f, right, h.toFloat())
            whiteKeys.add(PianoKey(i + 1, rect))

            if (i != 0 && i != 3 && i != 7) {
                left = (i - 1) * keyWidth +
                        0.5f * keyWidth +
                        0.25f * keyWidth
                right = i * keyWidth +
                        0.25f * keyWidth
                rect = RectF(left, 0f, right, 0.67f * h)
                blackKeys.add(PianoKey(count++, rect))
            }
        }
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        for (key in whiteKeys)
            canvas?.drawRect(key.rectF,
                if (key.isPressed) grayColor else whiteColor)

        for (i in 1 until numberOfKeys)
            canvas?.drawLine(i * keyWidth, 0f, i * keyWidth,
                keyHeight, blackColor)

        for (key in blackKeys)
            canvas?.drawRect(key.rectF,
                if (key.isPressed) grayColor else blackColor)
    }

    override fun performClick(): Boolean {
        super.performClick()
        return true
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        super.onTouchEvent(event)

        val action = event?.action
        val isPressed = action == MotionEvent.ACTION_DOWN ||
                action == MotionEvent.ACTION_MOVE

        if (action == MotionEvent.ACTION_UP)
            performClick()

        val pressedKeys: ArrayList<PianoKey> = ArrayList()

        for (i in 0 until event!!.pointerCount) {
            val x = event.getX(i)
            val y = event.getY(i)
            val key: PianoKey? = keyForCoordinates(x, y)
            if (key != null) {
                key.isPressed = isPressed
                pressedKeys.add(key)
            }
        }

        val keyboard: ArrayList<PianoKey> = ArrayList(whiteKeys)
        keyboard.addAll(blackKeys)

        for (key in keyboard) {
            if (pressedKeys.contains(key) && key.isPressed) {
                if (soundPlayer.isNotPlaying(key.sound)) {
                    soundPlayer.playNote(key.sound)
                    invalidate()
                }
            } else {
                soundPlayer.stopNote(key.sound)
                releasePianoKey(key)
            }
        }

        return true

    }

    private fun keyForCoordinates(x: Float, y: Float): PianoKey? {
        for (key in blackKeys)
            if (key.rectF.contains(x, y))
                return key

        for (key in whiteKeys)
            if (key.rectF.contains(x, y))
                return key

        return null
    }

    private fun releasePianoKey(key: PianoKey) {
        mHandler.postDelayed({
            key.isPressed = false
            mHandler.sendEmptyMessage(0)
        }, 100)
    }

}
